### PR TITLE
feat: add share reason in share requests

### PR DESCRIPTION
**Pull request by noah-paige**
### Feature or Bugfix
- Feature

### Detail
- Expose the Share Request Comments as Part of the Share View for the Approver to See 
- Allow Approver to Provide Comments when Rejecting a Share for the Requester to See

### Relates
- [#422 ]

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by noah-paige** [comment]
This PR adds the Share Request Purpose specified when requesting data access to the Share View so that the Approver of the share request is able to understand why the requester needs access
Screenshot 2023-06-06 at 11 45 54 AM
Screenshot 2023-06-06 at 11 47 42 AM

Additionally, this PR adds the ability of the approver to add the reject purpose when they reject a share request so the requester is able to see why the request was rejected
 
Screenshot 2023-06-06 at 11 48 56 AM
Screenshot 2023-06-06 at 11 49 16 AM

**Comment by dlpzx** [comment]
Hi @noah-paige thanks for the PR! So quick! I got one question, can users edit the reasons to request/reject?

File: backend/dataall/db/models/ShareObject.py
Comment (dlpzx) [reply backend/dataall/db/models/ShareObject.py:37]: Although it is handled by `{share.rejectPurpose || '-'} ` in frontend, so it is not needed

File: frontend/src/views/Shares/ShareView.js
Comment (dlpzx) [inline frontend/src/views/Shares/ShareView.js:153]: Since it is a free text field, users might add long descriptions that surpass the memory limits of an RDS cell. In the creation forms (CreateDataset...) we limit the description fields to 5000 chars (`frontend/src/views/Environments/EnvironmentCreateForm.js:375`). We can access the actual limits in the RDS database parameters, but I would say that in any case we need to do a check before writing to RDS

**Review by dlpzx — COMMENTED** [review]
in general it looks good

File: backend/dataall/db/models/ShareObject.py
Comment (noah-paige) [reply backend/dataall/db/models/ShareObject.py:37]: I am inclined to leave it as `null` rather than pass a default for these 2 columns

For `submit_share_object()` - the reason I include the line `share.rejectPurpose = ""` is for instances when a share is

1. Submitted by Requester
2. Rejected by Approver with Reason 1
3. Re-Submitted by Requester (changing Share Items requested)
4. Approver is able to reject again with new Reason 2

Maybe a better design would be to always allow the Approver to edit the reject reason and always allow the Requester to edit the request reason but I am not sure if that level of edit-ability is needed / would be used. What do you think?

File: frontend/src/views/Shares/ShareView.js
Comment (noah-paige) [reply frontend/src/views/Shares/ShareView.js:153]: Good call out - let me add a string length limit for the reasons provided

Comment (noah-paige) [reply frontend/src/views/Shares/ShareView.js:155]: Added this character limit

File: backend/dataall/db/models/ShareObject.py
Comment (noah-paige) [reply backend/dataall/db/models/ShareObject.py:37]: Added the edit ability in the latest commit where Requesters can edit requestPurpose and Approvers can edit rejectPurpose respectively

File: backend/migrations/versions/72b8a90b6ee8__share_request_purpose.py
Comment (dlpzx) [inline backend/migrations/versions/72b8a90b6ee8__share_request_purpose.py:28]: Important typo: `'reuestPurpose'` to `'requestPurpose'`

File: backend/dataall/db/api/share_object.py
Comment (dlpzx) [inline backend/dataall/db/api/share_object.py:617]: I see that you are checking that the caller belongs to the requesters in the resolvers, but why not decorating the method with `@has_resource_perm(permissions.SUBMIT_SHARE_OBJECT)`?

File: backend/migrations/versions/72b8a90b6ee8__share_request_purpose.py
Comment (noah-paige) [reply backend/migrations/versions/72b8a90b6ee8__share_request_purpose.py:28]: Great catch - let me push that fix now, thanks!!

File: backend/dataall/api/Objects/ShareObject/resolvers.py
Comment (noah-paige) [reply backend/dataall/api/Objects/ShareObject/resolvers.py:384]: Removing, thanks!

File: backend/dataall/api/Objects/ShareObject/resolvers.py
Comment (noah-paige) [reply backend/dataall/api/Objects/ShareObject/resolvers.py:404]: We could do it the same way as for updating share request purpose - I wanted to use the decorator `@has_resource_perm(permissions.REJECT_SHARE_OBJECT)` when possible which we can for updating reject purpose because Approver groups are the only group who are given the ability to reject/approve the share

**Comment by noah-paige** [comment]
I like this frontend design screenshot better - let me add that separate box and get a new commit up

**Comment by noah-paige** [comment]
Latest Commit:
- Removes print statement
- Fix Typo in Migration Script
- Change Share Object Comments Frontend View to a new box as shown in @dlpzx screenshot above

File: backend/dataall/db/api/share_object.py
Comment (dlpzx) [reply backend/dataall/db/api/share_object.py:617]: I see now why you implemented this with the "requesters" checker and the other with the permissions decorator checker! But then what I think is that the `permissions.SUBMIT_SHARE_OBJECT` is wrongly given to approvers. They are not able to see the `Submit` button in `frontend/src/views/Shares/ShareView.js` they should not be able to reach the api from the backend either. We need to modify `backend/dataall/db/permissions.py` and remove ` SUBMIT_SHARE_OBJECT,` from the list of permissions for the approver. In addition, we need to add some lines to the migration script to remove the existing permissions `SUBMIT_SHARE_OBJECT` 
```
SHARE_OBJECT_APPROVER = [
    ADD_ITEM,
    REMOVE_ITEM,
    SUBMIT_SHARE_OBJECT,
    APPROVE_SHARE_OBJECT,
    REJECT_SHARE_OBJECT,
    DELETE_SHARE_OBJECT,
    GET_SHARE_OBJECT,
    LIST_SHARED_ITEMS,
]
```
I am aware this is not the main purpose of this PR, but we could add this fix to the PR as it also includes a migration script.

Comment (noah-paige) [reply backend/dataall/db/api/share_object.py:617]: working on the revisions to the permission checker and migration script now :)

File: backend/migrations/versions/72b8a90b6ee8__share_request_purpose.py
Comment (dlpzx) [inline backend/migrations/versions/72b8a90b6ee8__share_request_purpose.py:82]: We are missing the back-filling when downgrading. I know, we hardly ever downgrade, but once I had a customer that needed to downgrade and it was crucial that migrations were both-sided

File: backend/dataall/db/api/share_object.py
Comment (dlpzx) [inline backend/dataall/db/api/share_object.py:1067]: good catch!

File: frontend/src/views/Shares/ShareView.js
Comment (dlpzx) [inline frontend/src/views/Shares/ShareView.js:714]: If I am picky and push this a bit by adding super long reasons, if the users do not start new lines (which works fine) they will lose part of their messages in the frontend

image

File: tests/api/conftest.py
Comment (dlpzx) [inline tests/api/conftest.py:487]: Should we remove this comment?

**Review by dlpzx — COMMENTED** [review]
final final details

File: tests/api/conftest.py
Comment (noah-paige) [reply tests/api/conftest.py:487]: Yes - let me clean this up

File: frontend/src/views/Shares/ShareView.js
Comment (noah-paige) [reply frontend/src/views/Shares/ShareView.js:714]: Good call out - added to the frontend `sx={{ wordBreak: "break-word" }}` and tested to ensure words can be broken and wrapped to new lines for both reject and request Purpose

File: backend/dataall/api/Objects/ShareObject/resolvers.py
Comment (noah-paige) [reply backend/dataall/api/Objects/ShareObject/resolvers.py:235]: You are absolutely correct - I switched to check `source.groupUri in context.groups` for the `resolve_user_role()` and tested locally the fix

File: backend/migrations/versions/72b8a90b6ee8__share_request_purpose.py
Comment (noah-paige) [reply backend/migrations/versions/72b8a90b6ee8__share_request_purpose.py:90]: I just updated the PR with code for downgrading and did tests on both upgrade and downgrade again on the Aurora DB 

Validated the resource permissions and columns got upgraded and downgraded as expected. @dlpzx - Think we are good here if you want to do one last final review

File: frontend/src/views/Shares/ShareView.js
Comment (dlpzx) [inline frontend/src/views/Shares/ShareView.js:768]: Nice

**Review by dlpzx — APPROVED** [review]
Looks good, thanks for including so many small fixes in this PR. I also re-tested the migration up and down an it is 🥇

**Pull request merged**
